File: frontend/lib/project-service.ts
import fs from 'fs';
import path from 'path';
import { simpleGit } from 'simple-git';
import { projectDb, Project, ProjectCreate } from './db';
import { dockerService } from './docker';
import { nginxService } from './nginx';

const PROJECTS_BASE_PATH = process.env.PROJECTS_BASE_PATH || path.join(process.cwd(), 'data', 'projects');
const PORT_RANGE_START = parseInt(process.env.PORT_RANGE_START || '10000');
const PORT_RANGE_END = parseInt(process.env.PORT_RANGE_END || '20000');

// Ensure projects directory exists
if (!fs.existsSync(PROJECTS_BASE_PATH)) {
  fs.mkdirSync(PROJECTS_BASE_PATH, { recursive: true });
}

function getAvailablePort(): number {
  const usedPorts = projectDb.getUsedPorts();
  for (let i = 0; i < 100; i++) {
    const port = Math.floor(Math.random() * (PORT_RANGE_END - PORT_RANGE_START)) + PORT_RANGE_START;
    if (!usedPorts.includes(port)) {
      return port;
    }
  }
  throw new Error('No available ports');
}

async function prepareProjectDirectory(project: Project): Promise<string> {
  const projectDir = path.join(PROJECTS_BASE_PATH, project.name);

  // Remove existing directory
  if (fs.existsSync(projectDir)) {
    fs.rmSync(projectDir, { recursive: true, force: true });
  }

  fs.mkdirSync(projectDir, { recursive: true });

  if (project.source_type === 'github') {
    // Clone from GitHub
    const git = simpleGit();
    await git.clone(project.source_path, projectDir);
  } else {
    // Copy from local path
    if (!fs.existsSync(project.source_path)) {
      throw new Error(`Source path does not exist: ${project.source_path}`);
    }

    if (fs.statSync(project.source_path).isDirectory()) {
      fs.cpSync(project.source_path, projectDir, { recursive: true });
    } else {
      fs.copyFileSync(project.source_path, path.join(projectDir, path.basename(project.source_path)));
    }
  }

  return projectDir;
}

function createDockerfile(projectDir: string, launchCommand?: string): string {
  const dockerfilePath = path.join(projectDir, 'Dockerfile.generated');

  let content = '';

  // Check for Node.js project
  if (fs.existsSync(path.join(projectDir, 'package.json'))) {
    content = `FROM node:18-alpine
WORKDIR /app
COPY package*.json ./
RUN npm install
COPY . .
EXPOSE 80
CMD ${launchCommand || '["npm", "start"]'}
`;
  }
  // Check for Python project
  else if (fs.existsSync(path.join(projectDir, 'requirements.txt'))) {
    content = `FROM python:3.11-slim
WORKDIR /app
COPY requirements.txt .
RUN pip install --no-cache-dir -r requirements.txt
COPY . .
EXPOSE 80
CMD ${launchCommand || '["python", "app.py"]'}
`;
  }
  // Check for Go project
  else if (fs.existsSync(path.join(projectDir, 'go.mod'))) {
    content = `FROM golang:1.21-alpine AS builder
WORKDIR /app
COPY go.* ./
RUN go mod download
COPY . .
RUN go build -o main .

FROM alpine:latest
WORKDIR /app
COPY --from=builder /app/main .
EXPOSE 80
CMD ${launchCommand || '["./main"]'}
`;
  }
  // Generic
  else {
    content = `FROM ubuntu:22.04
WORKDIR /app
COPY . .
EXPOSE 80
CMD ${launchCommand || '["bash"]'}
`;
  }

  fs.writeFileSync(dockerfilePath, content);
  return dockerfilePath;
}

export const projectService = {
  async createProject(data: ProjectCreate): Promise<Project> {
    // Check if name exists
    const existing = projectDb.getByName(data.name);
    if (existing) {
      throw new Error(`Project with name '${data.name}' already exists`);
    }

    // Create project record
    const project = projectDb.create(data);

    // Allocate port for serverside projects
    if (project.type === 'serverside') {
      const port = getAvailablePort();
      projectDb.update(project.id, { port });
    }

    return projectDb.getById(project.id)!;
  },

  async deployProject(projectId: number): Promise<Project> {
    const project = projectDb.getById(projectId);
    if (!project) {
      throw new Error('Project not found');
    }

    try {
      projectDb.update(projectId, { status: 'building' });

      // Prepare project directory
      const projectDir = await prepareProjectDirectory(project);

      if (project.type === 'static') {
        // Configure Nginx for static site
        nginxService.createStaticConfig(project.subdomain, projectDir);
        projectDb.update(projectId, { status: 'running' });
      } else {
        // Serverside project
        let dockerfilePath = path.join(projectDir, 'Dockerfile');
        if (!fs.existsSync(dockerfilePath)) {
          dockerfilePath = createDockerfile(projectDir, project.launch_command || undefined);
        }

        // Build image
        const imageTag = `paas-${project.name}:latest`;
        const buildSuccess = await dockerService.buildImage(
          projectDir,
          path.basename(dockerfilePath),
          imageTag
        );

        if (!buildSuccess) {
          throw new Error('Failed to build Docker image');
        }

        // Start container
        const containerId = await dockerService.createAndStartContainer({
          image: imageTag,
          name: `paas-${project.name}`,
          port: project.port!
        });

        if (!containerId) {
          throw new Error('Failed to start container');
        }

        projectDb.update(projectId, { container_id: containerId, dockerfile_path: dockerfilePath });

        // Configure Nginx reverse proxy
        nginxService.createProxyConfig(project.subdomain, project.port!);
        projectDb.update(projectId, { status: 'running' });
      }

      return projectDb.getById(projectId)!;
    } catch (error: any) {
      projectDb.update(projectId, {
        status: 'failed',
        error_message: error.message
      });
      throw error;
    }
  },

  async startProject(projectId: number): Promise<Project> {
    const project = projectDb.getById(projectId);
    if (!project) {
      throw new Error('Project not found');
    }

    if (project.status === 'running') {
      return project;
    }

    if (project.type === 'serverside' && project.container_id) {
      const success = await dockerService.startContainer(project.container_id);
      if (!success) {
        throw new Error('Failed to start container');
      }
    }

    projectDb.update(projectId, { status: 'running' });
    return projectDb.getById(projectId)!;
  },

  async stopProject(projectId: number): Promise<Project> {
    const project = projectDb.getById(projectId);
    if (!project) {
      throw new Error('Project not found');
    }

    if (project.type === 'serverside' && project.container_id) {
      await dockerService.stopContainer(project.container_id);
    }

    projectDb.update(projectId, { status: 'stopped' });
    return projectDb.getById(projectId)!;
  },

  async deleteProject(projectId: number): Promise<boolean> {
    const project = projectDb.getById(projectId);
    if (!project) {
      throw new Error('Project not found');
    }

    // Remove container
    if (project.container_id) {
      await dockerService.removeContainer(project.container_id);
    }

    // Remove Nginx config
    nginxService.removeConfig(project.subdomain);

    // Remove project directory
    const projectDir = path.join(PROJECTS_BASE_PATH, project.name);
    if (fs.existsSync(projectDir)) {
      fs.rmSync(projectDir, { recursive: true, force: true });
    }

    // Delete from database
    return projectDb.delete(projectId);
  },

  async getProjectLogs(projectId: number, tail: number = 100): Promise<string> {
    const project = projectDb.getById(projectId);
    if (!project) {
      throw new Error('Project not found');
    }

    if (project.type === 'serverside' && project.container_id) {
      return await dockerService.getContainerLogs(project.container_id, tail);
    }

    return 'No logs available for static projects';
  }
};

File: backend/services/project_service.py
import os
import shutil
import logging
import random
from typing import Optional, List
from sqlalchemy.orm import Session
from git import Repo, GitCommandError
from models import Project, ProjectType, ProjectStatus
from schemas import ProjectCreate, ProjectUpdate
from services.docker_service import DockerService
from services.nginx_service import NginxService
from config import settings

logger = logging.getLogger(__name__)


class ProjectService:
    def __init__(self):
        self.docker_service = DockerService()
        self.nginx_service = NginxService()
        self.projects_base_path = settings.projects_base_path
        os.makedirs(self.projects_base_path, exist_ok=True)

    def _get_available_port(self, db: Session) -> int:
        """Find an available port for a new project"""
        used_ports = {p.port for p in db.query(Project).filter(Project.port.isnot(None)).all()}
        
        for _ in range(100):  # Try 100 times
            port = random.randint(settings.port_range_start, settings.port_range_end)
            if port not in used_ports:
                return port
        
        raise Exception("No available ports found")

    def _prepare_project_directory(self, project: Project) -> str:
        """Prepare project directory from source"""
        project_dir = os.path.join(self.projects_base_path, project.name)
        
        # Remove existing directory if it exists
        if os.path.exists(project_dir):
            shutil.rmtree(project_dir)
        
        os.makedirs(project_dir, exist_ok=True)
        
        if project.source_type == "github":
            # Clone from GitHub
            try:
                logger.info(f"Cloning {project.source_path} to {project_dir}")
                Repo.clone_from(project.source_path, project_dir)
                logger.info(f"Successfully cloned repository")
            except GitCommandError as e:
                logger.error(f"Failed to clone repository: {e}")
                raise Exception(f"Failed to clone repository: {e}")
        else:
            # Copy from local path
            source_path = project.source_path
            if not os.path.exists(source_path):
                raise Exception(f"Source path does not exist: {source_path}")
            
            logger.info(f"Copying from {source_path} to {project_dir}")
            if os.path.isdir(source_path):
                shutil.copytree(source_path, project_dir, dirs_exist_ok=True)
            else:
                shutil.copy2(source_path, project_dir)
        
        return project_dir

    def _create_dockerfile(self, project_dir: str, launch_command: Optional[str]) -> str:
        """Create a Dockerfile for the project"""
        dockerfile_path = os.path.join(project_dir, "Dockerfile.generated")
        
        # Check if package.json exists (Node.js project)
        if os.path.exists(os.path.join(project_dir, "package.json")):
            dockerfile_content = """FROM node:18-alpine

WORKDIR /app

COPY package*.json ./
RUN npm install

COPY . .

EXPOSE 80

CMD ["npm", "start"]
"""
            if launch_command:
                dockerfile_content = dockerfile_content.replace('CMD ["npm", "start"]', f'CMD {launch_command}')
        
        # Check if requirements.txt exists (Python project)
        elif os.path.exists(os.path.join(project_dir, "requirements.txt")):
            dockerfile_content = """FROM python:3.11-slim

WORKDIR /app

COPY requirements.txt .
RUN pip install --no-cache-dir -r requirements.txt

COPY . .

EXPOSE 80

CMD ["python", "app.py"]
"""
            if launch_command:
                dockerfile_content = dockerfile_content.replace('CMD ["python", "app.py"]', f'CMD {launch_command}')
        
        # Check if go.mod exists (Go project)
        elif os.path.exists(os.path.join(project_dir, "go.mod")):
            dockerfile_content = """FROM golang:1.21-alpine AS builder

WORKDIR /app

COPY go.* ./
RUN go mod download

COPY . .
RUN go build -o main .

FROM alpine:latest
WORKDIR /app
COPY --from=builder /app/main .

EXPOSE 80

CMD ["./main"]
"""
            if launch_command:
                dockerfile_content = dockerfile_content.replace('CMD ["./main"]', f'CMD {launch_command}')
        
        # Generic Dockerfile
        else:
            dockerfile_content = f"""FROM ubuntu:22.04

WORKDIR /app

COPY . .

EXPOSE 80

CMD {launch_command if launch_command else '["bash"]'}
"""
        
        with open(dockerfile_path, 'w') as f:
            f.write(dockerfile_content)
        
        logger.info(f"Created Dockerfile at {dockerfile_path}")
        return dockerfile_path

    def create_project(self, db: Session, project_data: ProjectCreate) -> Project:
        """Create a new project"""
        # Check if project name already exists
        existing = db.query(Project).filter(Project.name == project_data.name).first()
        if existing:
            raise Exception(f"Project with name '{project_data.name}' already exists")
        
        # Create project record
        project = Project(
            name=project_data.name,
            type=project_data.type,
            source_type=project_data.source_type,
            source_path=project_data.source_path,
            subdomain=project_data.name,
            launch_command=project_data.launch_command,
            description=project_data.description,
            status=ProjectStatus.STOPPED
        )
        
        # Allocate port for serverside projects
        if project.type == ProjectType.SERVERSIDE:
            project.port = self._get_available_port(db)
        
        db.add(project)
        db.commit()
        db.refresh(project)
        
        logger.info(f"Created project: {project.name}")
        return project

    def deploy_project(self, db: Session, project_id: int) -> Project:
        """Deploy a project"""
        project = db.query(Project).filter(Project.id == project_id).first()
        if not project:
            raise Exception("Project not found")
        
        try:
            project.status = ProjectStatus.BUILDING
            project.error_message = None
            db.commit()
            
            # Prepare project directory
            project_dir = self._prepare_project_directory(project)
            
            if project.type == ProjectType.STATIC:
                # Configure Nginx for static site
                self.nginx_service.create_static_config(project.subdomain, project_dir)
                project.status = ProjectStatus.RUNNING
                logger.info(f"Deployed static project: {project.name}")
            
            else:  # SERVERSIDE
                # Check if Dockerfile exists
                dockerfile_path = os.path.join(project_dir, "Dockerfile")
                if not os.path.exists(dockerfile_path):
                    # Create Dockerfile
                    dockerfile_path = self._create_dockerfile(project_dir, project.launch_command)
                
                project.dockerfile_path = dockerfile_path
                
                # Build Docker image
                image_tag = f"paas-{project.name}:latest"
                if not self.docker_service.build_image(
                    os.path.basename(dockerfile_path),
                    image_tag,
                    project_dir
                ):
                    raise Exception("Failed to build Docker image")
                
                # Start container
                container_id = self.docker_service.create_and_start_container(
                    image_tag,
                    f"paas-{project.name}",
                    project.port
                )
                
                if not container_id:
                    raise Exception("Failed to start container")
                
                project.container_id = container_id
                
                # Configure Nginx reverse proxy
                self.nginx_service.create_proxy_config(project.subdomain, project.port)
                
                project.status = ProjectStatus.RUNNING
                logger.info(f"Deployed serverside project: {project.name}")
            
            db.commit()
            db.refresh(project)
            return project
        
        except Exception as e:
            logger.error(f"Failed to deploy project: {e}")
            project.status = ProjectStatus.FAILED
            project.error_message = str(e)
            db.commit()
            raise

    def start_project(self, db: Session, project_id: int) -> Project:
        """Start a stopped project"""
        project = db.query(Project).filter(Project.id == project_id).first()
        if not project:
            raise Exception("Project not found")
        
        if project.status == ProjectStatus.RUNNING:
            return project
        
        try:
            if project.type == ProjectType.SERVERSIDE and project.container_id:
                if self.docker_service.start_container(project.container_id):
                    project.status = ProjectStatus.RUNNING
                    project.error_message = None
                else:
                    raise Exception("Failed to start container")
            else:
                # For static projects, just update status
                project.status = ProjectStatus.RUNNING
                project.error_message = None
            
            db.commit()
            db.refresh(project)
            logger.info(f"Started project: {project.name}")
            return project
        
        except Exception as e:
            logger.error(f"Failed to start project: {e}")
            project.error_message = str(e)
            db.commit()
            raise

    def stop_project(self, db: Session, project_id: int) -> Project:
        """Stop a running project"""
        project = db.query(Project).filter(Project.id == project_id).first()
        if not project:
            raise Exception("Project not found")
        
        if project.status == ProjectStatus.STOPPED:
            return project
        
        try:
            if project.type == ProjectType.SERVERSIDE and project.container_id:
                self.docker_service.stop_container(project.container_id)
            
            project.status = ProjectStatus.STOPPED
            db.commit()
            db.refresh(project)
            logger.info(f"Stopped project: {project.name}")
            return project
        
        except Exception as e:
            logger.error(f"Failed to stop project: {e}")
            raise

    def delete_project(self, db: Session, project_id: int) -> bool:
        """Delete a project"""
        project = db.query(Project).filter(Project.id == project_id).first()
        if not project:
            raise Exception("Project not found")
        
        try:
            # Stop and remove container if it exists
            if project.container_id:
                self.docker_service.remove_container(project.container_id)
            
            # Remove Nginx config
            self.nginx_service.remove_config(project.subdomain)
            
            # Remove project directory
            project_dir = os.path.join(self.projects_base_path, project.name)
            if os.path.exists(project_dir):
                shutil.rmtree(project_dir)
            
            # Delete from database
            db.delete(project)
            db.commit()
            
            logger.info(f"Deleted project: {project.name}")
            return True
        
        except Exception as e:
            logger.error(f"Failed to delete project: {e}")
            raise

    def get_project(self, db: Session, project_id: int) -> Optional[Project]:
        """Get a project by ID"""
        return db.query(Project).filter(Project.id == project_id).first()

    def get_project_by_name(self, db: Session, name: str) -> Optional[Project]:
        """Get a project by name"""
        return db.query(Project).filter(Project.name == name).first()

    def list_projects(self, db: Session) -> List[Project]:
        """List all projects"""
        return db.query(Project).all()

    def update_project(self, db: Session, project_id: int, update_data: ProjectUpdate) -> Project:
        """Update project metadata"""
        project = db.query(Project).filter(Project.id == project_id).first()
        if not project:
            raise Exception("Project not found")
        
        if update_data.description is not None:
            project.description = update_data.description
        if update_data.launch_command is not None:
            project.launch_command = update_data.launch_command
        
        db.commit()
        db.refresh(project)
        return project

    def get_project_logs(self, db: Session, project_id: int, tail: int = 100) -> Optional[str]:
        """Get logs for a serverside project"""
        project = db.query(Project).filter(Project.id == project_id).first()
        if not project:
            raise Exception("Project not found")
        
        if project.type == ProjectType.SERVERSIDE and project.container_id:
            return self.docker_service.get_container_logs(project.container_id, tail)
        
        return None
